src/aead/quic.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright 2018 Brian Smith.
+//
+// Permission to use, copy, modify, and/or distribute this software for any
+// purpose with or without fee is hereby granted, provided that the above
+// copyright notice and this permission notice appear in all copies.
+//
+// THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHORS DISCLAIM ALL WARRANTIES
+// WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
+// SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+// WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION
+// OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF OR IN
+// CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+
+//! QUIC Header Protection.
+//!
+//! See draft-ietf-quic-tls.
+
+use crate::{
+    aead::{aes, block},
+    cpu, error,
+    polyfill::convert::*,
+};
+
+/// A key for generating QUIC Header Protection masks.
+pub struct HeaderProtectionKey {
+    inner: KeyInner,
+    algorithm: &'static Algorithm,
+}
+
+#[allow(variant_size_differences)]
+enum KeyInner {
+    Aes(aes::Key),
+}
+
+impl HeaderProtectionKey {
+    /// Create a new header protection key.
+    ///
+    /// `key_bytes` must be exactly `algorithm.key_len` bytes long.
+    pub fn new(
+        algorithm: &'static Algorithm, key_bytes: &[u8],
+    ) -> Result<Self, error::Unspecified> {
+        cpu::cache_detected_features();
+        Ok(HeaderProtectionKey {
+            inner: (algorithm.init)(key_bytes)?,
+            algorithm,
+        })
+    }
+
+    /// Generate a new QUIC Header Protection mask.
+    ///
+    /// `sample` must be exactly `SAMPLE_LEN` bytes long.
+    pub fn new_mask(&self, sample: &[u8]) -> Result<[u8; SAMPLE_LEN], error::Unspecified> {
+        let sample = <&[u8; SAMPLE_LEN]>::try_from_(sample).unwrap();
+        let sample = block::Block::from(sample);
+        let block = (self.algorithm.new_mask)(&self.inner, sample);
+
+        // FIXME: turn Block into [u8; 16] without copying?
```

Comment:
If it actually helps performance, we could return a `Block` (and make `Block` public) to allow users to avoid the copy. However, keep in mind that a 16-byte copy should be just one or a few vector instructions on most targets.

Reply:
I'm fine leaving the copy, I just didn't know whether there was something already available to do the conversion. Anyway, I removed the comment.